Disable link suggestions when value is URL

danielbachhuber opened the pull request:
## Description

1. Disables link suggestions when `value` looks like a URL.
2. Disables `onKeyDown` handling when link suggestions are loading.

## How Has This Been Tested?

I've manually tested this locally.

## Screenshots (jpeg or gifs if applicable):

![urlsuggestions](https://user-images.githubusercontent.com/36432/35305727-40e9d166-004f-11e8-8167-ad94854099a3.gif)

See #4053

danielbachhuber committed:
Disable link suggestions when value is URL

danielbachhuber commented on blocks/url-input/index.js:
```diff
@@ -61,6 +61,17 @@ class UrlInput extends Component {
 		}
 
+		// Disable suggestions for URLs
+		if ( 0 === value.indexOf( 'http://' ) || 0 === value.indexOf( 'https://' ) ) {
```
Do we have an `isURL()` or similar helper for all protocols?

jaswrks replied:
Maybe use a regex to catch `any://`

```js
if ( /^\w+:\/\//.test( value ) ) {
```

There's also the built-in URL module:
https://nodejs.org/docs/latest-v7.x/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost

Maybe a good place for other URL utilities:
https://github.com/WordPress/packages/blob/master/packages/url/src/index.js

danielbachhuber replied:
> Maybe use a regexp to catch `any://`

I think this would be more durable, if we don't want to introduce a dedicated utility.

danielbachhuber replied:
70dc202b5dbc369dd8799038e5839a28aaa80ece updates the check to use the regex our existing TinyMCE plugin uses.

danielbachhuber commented:
> How Has This Been Tested?

Are there any meaningful unit tests I could add?

aduth commented on blocks/url-input/index.js:
```diff
@@ -109,9 +120,9 @@ class UrlInput extends Component {
+		// If the suggestions are not shown or loading, we shouldn't handle the arrow keys
 		// We shouldn't preventDefault to allow block arrow keys navigation
-		if ( ! this.state.showSuggestions || ! this.state.posts.length ) {
+		if ( ! this.state.showSuggestions || ! this.state.posts.length || this.suggestionsRequest ) {
```
If there had been a previous request which has since completed, wouldn't `this.suggestionsRequest` still be truthy because it's never unset?

danielbachhuber replied:
If there had been a previous request which has since completed, wouldn't `this.suggestionsRequest` still be truthy because it's never unset?

Good question. I don't know off the top of my head.

danielbachhuber replied:
Updated in 7ab4af6a2e356f2dfc6c802b7bffe29c0d213e83

danielbachhuber commented:
@WordPress/gutenberg-core Could I get a list of the changes required for this to land? Thanks

aduth commented:
>>If there had been a previous request which has since completed, wouldn't `this.suggestionsRequest` still be truthy because it's never unset?
>
>Good question. I don't know off the top of my head.

I'm fairly certain the answer is yes, in which case this needs to be fixed.

danielbachhuber committed:
Use regex check used by WPLink TinyMCE plugin

danielbachhuber committed:
Load these variables into scope as constants

danielbachhuber commented:
This is ready for another review. Here's a GIF of the current behavior:

![linksearching](https://user-images.githubusercontent.com/36432/39497337-b48da0b6-4d57-11e8-8dbe-f9238246db8d.gif)

noisysocks commented:
This looks good to me 👍 

Looks like the build is failing due to a lint error:

```
[0] /home/travis/build/WordPress/gutenberg/blocks/url-input/index.js
[0]   66:34  error  Multiple spaces found before ')'  no-multi-spaces
```

gziolo commented:
Just saying that when you type `ht` it still tries to load suggestions. You can notice it in the recorded screencast. This is because of the following logic:
```js
// Show the suggestions after typing at least 3 characters
if ( value.length < 2 ) {
...
```
The comment left there, it lies :)
It looks like typing slowly `ht`, `htt` will trigger the search.

danielbachhuber committed:
Fix linting issue

danielbachhuber committed:
Correct the comment

danielbachhuber commented:
> Just saying that when you type `ht` it still tries to load suggestions. You can notice it in the recorded screencast.

Good point. This is apparently the behavior for Classic link search too:

![linksearchingclassic](https://user-images.githubusercontent.com/36432/39527836-5168c116-4dd7-11e8-8039-e94371321f43.gif)

> The comment left there, it lies :)

Fixed in 88b6f03e837e9f31bd053e90aee18566341029cc

> Looks like the build is failing due to a lint error:

Fixed in 149d550cd44fd0697efc22be46396f8f2a283419

gziolo reviewed (approved):
LGTM

iseulde commented:
> It looks like typing slowly ht, htt will trigger the search.

I think that's fine? You may be searching for an article about HTTPS? :)

iseulde commented on blocks/url-input/index.js:
```diff
@@ -62,6 +62,17 @@ class UrlInput extends Component {
 		}
 
+		// Disable suggestions for URLs
+		if ( /^https?:/.test( value ) ) {
```
Is it worth merging this with the previous block?

danielbachhuber replied:
> Is it worth merging this with the previous block?

Yes, I think that's worthwhile: e171100cb

danielbachhuber committed:
Consolidate duplicative statements
